src/main/scala/nl.knaw.dans.dd.prepub/DansDataVaultMetadataBlockMapper.scala

```diff
@@ -19,43 +19,89 @@ import java.net.URI
-
-    // TODO: How find out if a bagId that is found here is carried over from the previous version or a bagId minted by the SWORD service?
-    //       Compare with bagId of previous version if exists
-    //       if no previous version: filled in bagId means SWORD filled it in
-    //       if previous version: new bagId means SWORD filled it in, the same bagId as in previous version means UI created new draft.
-    fields.append(PrimitiveSingleValueField("dansBagId", b.getOrElse(mintBagId())))
-    fields.append(PrimitiveSingleValueField("dansNbn", n.getOrElse(mintUrnNbn())))
+    fields.append(PrimitiveSingleValueField("dansBagId", bagId))
+    fields.append(PrimitiveSingleValueField("dansNbn", urn))
     o.foreach(b => fields.append(PrimitiveSingleValueField("dansOtherId", b)))
     ov.foreach(b => fields.append(PrimitiveSingleValueField("dansOtherIdVersion", b)))
     st.foreach(b => fields.append(PrimitiveSingleValueField("dansSwordToken", b)))
     FieldList(fields.toList)
   }
 
-  //TODO: add error handling
-  def mintUrnNbn(): String = {
-    Http(s"${pidGeneratorBaseUrl resolve "create"}?type=urn")
+  private def setBagId(bagIdOpt: Option[String], pid: String): Try[String] = bagIdOpt match {
+    case Some(bagId) => checkBagIdOrigin(bagId, pid)
+    case None => Success(mintBagId())
+  }
+
+  /**
+   *
+   * @param bagId the bagId found in the Vault metadatablock
+   * @param pid   the dataset pid
+   * @return
+   * if no previous version: filled in bagId means SWORD filled it in
+   * if previous version: new bagId means SWORD filled it in, the same bagId as in previous version means UI created new draft.
+   */
+  private def checkBagIdOrigin(bagId: String, pid: String): Try[String] = {
+    getBagIdOfPreviousVersion(pid)
+      .map(bagIdOpt => {
+        if (bagId.equals(bagIdOpt.getOrElse("")))
+          mintBagId()
+        else
+          bagId
+      })
+  }.recoverWith { case e: Exception => Failure(ExternalSystemCallException(s"Problem with Dataverse api: $e")) }
+
+  private def getBagIdOfPreviousVersion(pid: String): Try[Option[String]] = {
+    for {
+      response <- dataverse.dataset(pid).view(Version.LATEST_PUBLISHED)
+      dsv <- response.data
+      optBagId = dsv.metadataBlocks.get("dansDataVaultMetadata")
+        .flatMap(_.fields
+          .map(_.asInstanceOf[PrimitiveSingleValueField])
+          .find(_.typeName == "dansBagId"))
+        .map(_.value)
+    } yield optBagId
+  }
+
+  def mintUrnNbn(): Try[String] = Try {
+    Http(s"${ pidGeneratorBaseUrl resolve "create" }?type=urn")
       .method("POST")
-      .header("content-type", "*/*")
-      .header("accept", "*/*")
+      .header("Accept", "application/json")
       .asString.body
-  }
+  }.recoverWith { case e: Exception => Failure(ExternalSystemCallException(s"Problem with pid-generator service: $e")) }
```

Comment:
Keep the stack trace? Same at line 85.
```suggestion
  }.recoverWith { case e: Exception => Failure(ExternalSystemCallException(s"Problem with pid-generator service: $e",e)) }
```